frame/balances/src/lib.rs

```diff
@@ -603,1601 +467,606 @@ pub mod pallet {
-	pub struct PositiveImbalance<T: Config<I>, I: 'static = ()>(T::Balance);
-
-	impl<T: Config<I>, I: 'static> PositiveImbalance<T, I> {
-		/// Create a new positive imbalance from a balance.
-		pub fn new(amount: T::Balance) -> Self {
-			PositiveImbalance(amount)
 		}
-	}
 
-	/// Opaque, move-only struct with private fields that serves as a token denoting that
-	/// funds have been destroyed without any equal and opposite accounting.
-	#[must_use]
-	#[derive(RuntimeDebug, PartialEq, Eq)]
-	pub struct NegativeImbalance<T: Config<I>, I: 'static = ()>(T::Balance);
-
-	impl<T: Config<I>, I: 'static> NegativeImbalance<T, I> {
-		/// Create a new negative imbalance from a balance.
-		pub fn new(amount: T::Balance) -> Self {
-			NegativeImbalance(amount)
-		}
-	}
-
-	impl<T: Config<I>, I: 'static> TryDrop for PositiveImbalance<T, I> {
-		fn try_drop(self) -> result::Result<(), Self> {
-			self.drop_zero()
-		}
-	}
-
-	impl<T: Config<I>, I: 'static> Default for PositiveImbalance<T, I> {
-		fn default() -> Self {
-			Self::zero()
-		}
-	}
-
-	impl<T: Config<I>, I: 'static> Imbalance<T::Balance> for PositiveImbalance<T, I> {
-		type Opposite = NegativeImbalance<T, I>;
-
-		fn zero() -> Self {
-			Self(Zero::zero())
-		}
-		fn drop_zero(self) -> result::Result<(), Self> {
-			if self.0.is_zero() {
-				Ok(())
-			} else {
-				Err(self)
+		/// Upgrade a specified account.
+		///
+		/// - `origin`: Must be `Signed`.
+		/// - `who`: The account to be upgraded.
+		///
+		/// This will waive the transaction fee if at least all but 10% of the accounts needed to
+		/// be upgraded. (We let some not have to be upgraded just in order to allow for the
+		/// possibililty of churn).
+		#[pallet::call_index(6)]
+		#[pallet::weight(T::WeightInfo::upgrade_accounts(who.len() as u32))]
+		pub fn upgrade_accounts(
+			origin: OriginFor<T>,
+			who: Vec<T::AccountId>,
+		) -> DispatchResultWithPostInfo {
+			ensure_signed(origin)?;
+			if who.is_empty() {
+				return Ok(Pays::Yes.into())
 			}
-		}
-		fn split(self, amount: T::Balance) -> (Self, Self) {
-			let first = self.0.min(amount);
-			let second = self.0 - first;
-
-			mem::forget(self);
-			(Self(first), Self(second))
-		}
-		fn merge(mut self, other: Self) -> Self {
-			self.0 = self.0.saturating_add(other.0);
-			mem::forget(other);
-
-			self
-		}
-		fn subsume(&mut self, other: Self) {
-			self.0 = self.0.saturating_add(other.0);
-			mem::forget(other);
-		}
-		fn offset(self, other: Self::Opposite) -> SameOrOther<Self, Self::Opposite> {
-			let (a, b) = (self.0, other.0);
-			mem::forget((self, other));
-
-			if a > b {
-				SameOrOther::Same(Self(a - b))
-			} else if b > a {
-				SameOrOther::Other(NegativeImbalance::new(b - a))
+			let mut upgrade_count = 0;
+			for i in &who {
+				let upgraded = Self::ensure_upgraded(i);
+				if upgraded {
+					upgrade_count.saturating_inc();
+				}
+			}
+			if if Perbill::from_rational(upgrade_count, who.len()) >= Perbill::from_percent(90) { {
```

Comment:
There was a mistake in the last commit that won't compile
```suggestion
			if Perbill::from_rational(upgrade_count, who.len()) >= Perbill::from_percent(90) {
```